### ISSUE TITLE
ocr performance 

### ISSUE THREAD

**Issue by annopackage** [QwenLM/Qwen2-VL#241]
hi, thanks for your great work. I found the results on docvqa and textvqa are much better than other open source models. Can you share any insights or instructions for improving model ocr performance?

**Comment by 99991** [comment]
I would also be interested in the best way to use Qwen2-VL for OCR. Its ability to read handwritten text are phenomenal, but it has two big issues for which I have not yet found a solution:

1. The model is lazy and outputs something like `![Example diagram](image.png)` instead of actually transcribing the diagram. I could of course employ prompt engineering strategies like threatening the model that I am going to kill a million kittens if it does not comply, which helps a bit, but is not a general solution and also feels unethical.
2. The model gets stuck in a repeating loop. This can be prevented to some degree by setting `repeat_penalty` in `GenerationConfig`, but if the image really contains repeated text, it will be destroyed. Also, larger `repeat_penalty` increases the risk that the model comments on the image instead of transcribing its content.

Here is an example of the model being lazy:

```markdown
# Example diagram

![Example diagram](image.png)

# Example table

| User | name | email | ip |
|------|------|-------|----|
| 1    | Daisy | daisy@example.com | 927.00.1 |
| 2    | Donald | donald@ducktown.duck | 8.8.8.8 |
| 3    | Goofy | goofy@example.com | 6.6.6.6 |

# Example text

Hello, world! The quick brown fox jumps over the lazy dog.[1]
```

# image.jpg

![image](https://github.com/user-attachments/assets/d9b8e60a-e584-4b15-8ba0-fe8724c31b61)

# Code demonstrating lazy model

```python
from PIL import Image
import torch
from transformers import Qwen2VLForConditionalGeneration, AutoTokenizer, AutoProcessor, GenerationConfig

model_name = "Qwen/Qwen2-VL-7B-Instruct"

model = Qwen2VLForConditionalGeneration.from_pretrained(model_name, torch_dtype=torch.float16, device_map="cuda")
processor = AutoProcessor.from_pretrained(model_name, device="cuda")
generation_config = GenerationConfig().from_pretrained(model_name, do_sample=False)

image = Image.open("image.jpg")

conversation = [
    # System prompt seems to be ignored. The model outputs markdown instead of latex, even though the system prompt should override what the user wants.
    {"role": "system", "content": "You are an OCR engine which takes an image and converts it to latex, even if the user asks for a different format."},
    {
        "role": "user",
        "content": [
            {"type": "image",},
            # Ignores command to write markdown directly
            {"type": "text", "text": "Transcribe everything in this image as markdown. Start directly with the markdown content. Do not write ```markdown or anything else first."},
        ],
    }
]


text_prompt = processor.apply_chat_template(conversation, add_generation_prompt=True)
inputs = processor(text=[text_prompt], images=[image], padding=True, return_tensors="pt")
inputs = inputs.to("cuda")

output_ids = model.generate(**inputs, max_new_tokens=512, generation_config=generation_config)

generated_ids = [output_ids[len(input_ids) :] for input_ids, output_ids in zip(inputs.input_ids, output_ids)]

text, = processor.batch_decode(generated_ids, skip_special_tokens=True, clean_up_tokenization_spaces=True)

print(text)
```

And here is an example for the model being repetitive. This problem seems to happen mostly with images of rather low quality. To be fair, the low quality image below does not really contain much that can be deciphered, but the model definitely should not get stuck in an finite loop.

```markdown
MSE
0
1
2
3

MSE
0
1
2

MSE
0
1
2
3

... [goes on forever like this]
```

# low_quality.jpg

![low_quality](https://github.com/user-attachments/assets/3c116369-866b-4151-a154-f840369090b2)

# Code demonstrating stuck LLM repeating the same thing until token limit is exhausted

```python
from PIL import Image
import torch
from transformers import Qwen2VLForConditionalGeneration, AutoTokenizer, AutoProcessor, GenerationConfig

model_name = "Qwen/Qwen2-VL-7B-Instruct"

model = Qwen2VLForConditionalGeneration.from_pretrained(model_name, torch_dtype=torch.float16, device_map="cuda")
processor = AutoProcessor.from_pretrained(model_name, device="cuda")

# larger values for repetition_penalty make the repetition bug go away, but OCR quality for other images is worse
generation_config = GenerationConfig().from_pretrained(model_name, repetition_penalty=1.0)

image = Image.open("low_quality.jpg")

conversation = [
    {
        "role": "user",
        "content": [
            {"type": "image",},
            {"type": "text", "text": "Transcribe everything as markdown. Do not use placeholders like [image here] or something like that."},
        ],
    }
]

text_prompt = processor.apply_chat_template(conversation, add_generation_prompt=True)
inputs = processor(text=[text_prompt], images=[image], padding=True, return_tensors="pt")
inputs = inputs.to("cuda")

output_ids = model.generate(**inputs, max_new_tokens=256, generation_config=generation_config)

generated_ids = [output_ids[len(input_ids) :] for input_ids, output_ids in zip(inputs.input_ids, output_ids)]

text, = processor.batch_decode(generated_ids, skip_special_tokens=True, clean_up_tokenization_spaces=True)

print(text)
```

**Comment by hengkuankian** [comment]
+1

**Comment by DemonDamon** [comment]
+1

**Comment by n00mkrad** [comment]
+1

repetition is a big problem

**Comment by n-kostadinov** [comment]
+1 tried different frequency_penalty values. Behaviour is inconsistent, and loops happen even with high values for frequency_penalty

**Comment by n00mkrad** [comment]
A somewhat reliable method is to rotate images that consistently give you repetitions. So, while not being very efficient, a brute-force way to get reliable OCR is to "multi-sample" a few times with a random rotation applied to the image.

**Comment by Holmes2002** [comment]
+1, I encounterd repetition problem therefore I used repetition as 1.2. However, performance was inconsistent, sometimes this issue appeared. Any solutions ?

**Comment by n-kostadinov** [comment]
hi, not exactly a solution here, yet it might be useful. In langchain we configured fallbacks where FREQUENCY_PENALTY gets higher whenever the model goes into infinite loop. with frequency_penalty=0.01 it never goes into infinite loop for our cases, yet it the result is shorter than expected. But at least we get a result...

**Comment by Holmes2002** [comment]
@n-kostadinov Thanks for your advice! I will try it

**Comment by 99991** [comment]
This issue has not been fixed as far as I can tell.

@ShuaiBai623 Could you elaborate why you closed this issue?

EDIT: It looks like they are just closing all issues.

**Comment by dinhngoc267** [comment]
Heyyy @99991,  

```
# Example diagram
 ![Example diagram](image.png)
```

Then do you know how to extract the bounding box of the image? or how to extract the image also? Should I do another object detection task, or it's able to do end to end?

I tried deepseek https://huggingface.co/spaces/merterbak/DeepSeek-OCR-Demo. It's able to do that. Thank you

Thank you a lot!

**Comment by 99991** [comment]
> Then do you know how to extract the bounding box of the image?

Maybe you can get bounding boxes by simply asking for them as JSON. See section ["5. 2D/3D Grounding"](https://qwen.ai/blog?id=99f0335c4ad9ff6153e517418d48535ab6d8afef&from=research.latest-advancements-list#:~:text=5.%202D/3D%20Grounding). 

Here are a few more options to try:

* [PaddleOCR](https://github.com/PaddlePaddle/PaddleOCR?tab=readme-ov-file#-quick-overview-of-execution-results)
* [dots.ocr](https://github.com/rednote-hilab/dots.ocr?tab=readme-ov-file#example-for-formula-document)
* [RapidOCR](https://github.com/RapidAI/RapidOCR?tab=readme-ov-file#-visualization)
* [Docling](https://github.com/docling-project/docling)
* [DocLayout-YOLO](https://github.com/opendatalab/DocLayout-YOLO)
* [MinerU](https://github.com/opendatalab/MinerU?tab=readme-ov-file#mineru)
* [MonkeyOCR](https://github.com/Yuliang-Liu/MonkeyOCR?tab=readme-ov-file#support-diverse-chinese-and-english-pdf-types)

And of course all the big players (Microsoft, Amazon, OpenAI, Anthropic, Mistral) have APIs for OCR. They are not free, but usually among the best.

Lastly, here are a few benchmarks for OCR with even more models:

* https://idp-leaderboard.org/
* https://github.com/opendatalab/OmniDocBench
